File: 2020/day06/puzzle6.py
file = open('puzzle6.in')

groups = []
current_group = { 'people_count': 0, 'answers': {} }

for line in file:
  if line.strip() == "":
    groups.append(current_group)
    current_group = { 'people_count': 0, 'answers': {} }
  else:
    current_group['people_count'] += 1;
    for char in line.strip():
      if char in current_group['answers']:
        current_group['answers'][char] += 1
      else: 
        current_group['answers'][char] = 1
groups.append(current_group)

part1_count = 0
part2_count = 0
for group in groups:
  part1_count += len(group['answers'])

  for question in group['answers']:
    if group['answers'][question] == group['people_count']:
      part2_count += 1

print(part1_count)
print(part2_count)

File: 2020/day06/puzzle6.ts
const input = await Deno.readTextFile("puzzle6.in");

type AnswerObject = { [question: string]: number };
type Group = { count: number; answers: AnswerObject };

function createAnswerObjects(answers: string[]): AnswerObject[] {
  return answers
    .map((answer) => Array.from(answer).map((question: string) => [question, 1]))
    .map((answer) => Object.fromEntries(answer));
}

function sumAnswers(answers: AnswerObject[]) {
  return {
    count: answers.length,
    answers: answers.reduce((sum, answerObject) => mergeAnswerObjects(answerObject, sum), {}),
  };
}

function mergeAnswerObjects(fromObject: AnswerObject, toObject: AnswerObject) {
  const mergedAnswers = { ...toObject };
  for (const question in fromObject) {
    if (mergedAnswers[question]) {
      mergedAnswers[question] += fromObject[question];
    } else {
      mergedAnswers[question] = fromObject[question];
    }
  }
  return mergedAnswers;
}

function sumPart1(sum: number, group: Group) {
  return sum + Object.values(group.answers).length;
}

function sumPart2(sum: number, group: Group) {
  return sum + Object.values(group.answers).filter((count: number) => count === group.count).length;
}

const groups = input
  .split(/\n\n/)
  .map((group: string) => group.split(/\n/))
  .map(createAnswerObjects)
  .map(sumAnswers);

console.log(groups.reduce(sumPart1, 0));
console.log(groups.reduce(sumPart2, 0));
